src/packageurl/contrib/url2purl.py

```diff
@@ -276,3 +294,100 @@ def build_sourceforge_purl(uri):
 @purl_router.route(cargo_pattern)
 def build_cargo_purl(uri):
     return purl_from_pattern('cargo', cargo_pattern, uri)
+
+
+@purl_router.route('https?://github.com/.*')
+def build_github_purl(uri):
+    """
+    Take `uri` as input 
+    For example:
+    https://github.com/nexB/scancode-toolkit/tree/develop/plugins/scancode-ctags-macosx_10_9_intel
+    """
+    data = base_parsing(uri)
+
+    if data:
+        segments = data['segments']
+        name = data['name']
+        namespace = data['namespace']
+        version = None
+        subpath = None
+
+        #https://github.com/TG1999/fetchcode/master
+        if len(segments) >= 3  and segments[2] != 'tree':
+            version = segments[2]
+            subpath = '/'.join(segments[3:])
+
+        #https://github.com/TG1999/fetchcode/tree/master
+        if len(segments) >= 4 and segments[2] == 'tree':
+            version = segments[3]
+            subpath = '/'.join(segments[4:])
+        
+        return PackageURL(type='github', namespace=namespace, name=name, version=version, subpath=subpath)
+
+    else:
+        return
+
+ 
+@purl_router.route('https?://bitbucket.org/.*')
+def build_bitbucket_purl(uri):
+    """
+    Take `uri` as input
+    For exmaple:
+    https://bitbucket.org/TG1999/first_repo/src/master
+    """
+    data = base_parsing(uri)
+
+    if data:
+        segments = data['segments']
+        name = data['name']
+        namespace = data['namespace']
+        version = None
+        subpath = None
+
+        #https://bitbucket.org/TG1999/first_repo/new_folder/
+        if len(segments) >= 3  and segments[2] != 'src':
+            version = segments[2]
+            subpath = '/'.join(segments[3:])
+
+        #https://bitbucket.org/TG1999/first_repo/src/master/new_folder/
+        if len(segments) >= 4 and segments[2] == 'src':
+            version = segments[3]
+            subpath = '/'.join(segments[4:])
+        
+        return PackageURL(type='bitbucket', namespace=namespace, name=name, version=version, subpath=subpath)
+
+    else:
+        return
+
+
+@purl_router.route('https?://gitlab.com/.*')
+def build_gitlab_purl(uri):
+    """
+    Take `uri` as input
+    For example:
+    https://gitlab.com/TG1999/firebase/-/tree/1a122122/views
+    """
+    data = base_parsing(uri)
+
+    if data:
+        segments = data['segments']
+        name = data['name']
+        namespace = data['namespace']
+        version = None
+        subpath = None
+
+        #https://gitlab.com/TG1999/firebase/master 
+        if (len(segments) >= 3) and segments[2] != '-' and segments[2] != 'tree' :
```

Comment:
These multiple list element references are very hard to read. Can we give these some proper variable names?

Reply:
Yeah sure, how does `seperator` sounds

Reply:
That is fine if that is what `segments[2]` represents. 

The point here is that an outside user of this code should be able to read the code and understand what is going on without having to parse out what each element of an list is.